contracts/price_oracle/price_feed/NapierPriceFeed.sol

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.26;
+
+import {SafeCast} from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+import {IPriceFeed} from "./IPriceFeed.sol";
+import {IPriceOracleMiddleware} from "../IPriceOracleMiddleware.sol";
+import {AggregatorV3Interface} from "../ext/AggregatorV3Interface.sol";
+
+/// @notice Standard interface for {TokiTWAPChainlinkOracle} and {TokiLinearChainlinkOracle}
+interface ITokiChainlinkCompatOracle is AggregatorV3Interface {
+    /// @notice Returns the immutable arguments of the Toki Chainlink compatible oracle
+    /// @dev Linear discount oracle supports only PT as base asset
+    /// @return liquidityToken Address of the Napier liquidity token (Toki pool token)
+    /// @return base Address of the base asset (PT or LP)
+    /// @return quote Address of the pricing asset (asset or underlying)
+    function parseImmutableArgs() external view returns (address liquidityToken, address base, address quote, uint256);
+}
+
+/// @title Price feed for Napier v2 Principal Tokens and LP tokens
+/// @notice Provides USD price data for Napier PT tokens using the Toki chainlink compatible oracle
+/// @dev Implementation notes:
+/// - Expects the provided Toki chainlink compatible oracle to be pre-initialized and populated for the pool
+/// - PriceOracleMiddleware must have a source configured for the chosen pricing asset (asset or underlying),
+///   as determined by the Toki oracle's immutable configuration
+contract NapierPriceFeed is IPriceFeed {
+    using SafeCast for *;
+
+    uint8 public constant TOKI_CHAINLINK_ORACLE_DECIMALS = 18;
+
+    /// @notice Address of the Napier Toki Chainlink AggregatorV3Interface compatible oracle
+    ITokiChainlinkCompatOracle public immutable TOKI_CHAINLINK_ORACLE;
+
+    /// @notice Address of the Napier liquidity token (Toki pool token)
+    address public immutable LIQUIDITY_TOKEN;
+
+    /// @notice Address of the price oracle middleware expected to supply USD prices
+    address public immutable PRICE_MIDDLEWARE;
```

Comment:
Consider instead of setting this here to call it on msg.sender(IPriceOracleMiddleware(PRICE_MIDDLEWARE)), the advantage of such a solution is that then the same feed can be used across different vaults and then each ATOMISTA (user with the appropriate role) can determine how the QUOTE will be priced, and then all the power remains under the curator of the one managing the vault. An example is in ERC4626PriceFeed.